File: Python/ExampleTests/common/cart_input_typed.py
from . import json_util as _json
from .cart_input_string import CartInputString

class CartInputTyped:
    def __init__(self, total_items: str = '', shipping: str = '', discount: str = '', total_price: str = '', notes: str = ''):
        self.total_items = total_items
        self.shipping = shipping
        self.discount = discount
        self.total_price = total_price
        self.notes = notes

    @classmethod
    def from_string_obj(cls, s: CartInputString) -> 'CartInputTyped':
        return cls(
            s.total_items,
            s.shipping,
            s.discount,
            s.total_price,
            s.notes
        )

    def to_string_obj(self) -> CartInputString:
        return CartInputString(
            str(self.total_items),
            str(self.shipping),
            str(self.discount),
            str(self.total_price),
            str(self.notes)
        )

    @staticmethod
    def to_string_list(items) -> list:
        return [t.to_string_obj() for t in items]

    @staticmethod
    def from_string_list(items) -> list:
        return [CartInputTyped.from_string_obj(s) for s in items]

    def to_json_value(self) -> dict:
        return {
            'TotalItems': self.total_items,
            'Shipping': self.shipping,
            'Discount': self.discount,
            'Total Price': self.total_price,
            'Notes': self.notes,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'CartInputTyped':
        return cls(
            _json.as_str(_json.require(m, 'TotalItems'), 'TotalItems'),
            _json.as_str(_json.require(m, 'Shipping'), 'Shipping'),
            _json.as_str(_json.require(m, 'Discount'), 'Discount'),
            _json.as_str(_json.require(m, 'Total Price'), 'Total Price'),
            _json.as_str(_json.require(m, 'Notes'), 'Notes')
        )

    @classmethod
    def from_json(cls, text: str) -> 'CartInputTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'CartInputTyped')
        return [cls.from_json_value(e) for e in raw]

    def __str__(self):
        return (f'TotalItems={self.total_items}' + ', ' +
                f'Shipping={self.shipping}' + ', ' +
                f'Discount={self.discount}' + ', ' +
                f'Total Price={self.total_price}' + ', ' +
                f'Notes={self.notes}')

    def _key(self):
        return (self.total_items, self.shipping, self.discount, self.total_price, self.notes)

    def __eq__(self, other):
        if not isinstance(other, CartInputTyped):
            return NotImplemented
        return self._key() == other._key()

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'CartInputTyped({self})'

File: JavaScript/ExampleTests/common/CartInputTyped.js
import { CartInputString } from "./CartInputString.js";
import * as _json from "./json.js";

export class CartInputTyped {
  constructor(totalItems = "", shipping = "", discount = "", totalPrice = "", notes = "") {
    this.totalItems = totalItems;
    this.shipping = shipping;
    this.discount = discount;
    this.totalPrice = totalPrice;
    this.notes = notes;
  }

  static fromStringObj(s) {
    return new CartInputTyped(
      s.totalItems,
      s.shipping,
      s.discount,
      s.totalPrice,
      s.notes
    );
  }

  toStringObj() {
    return new CartInputString(
      String(this.totalItems),
      String(this.shipping),
      String(this.discount),
      String(this.totalPrice),
      String(this.notes)
    );
  }

  static toStringList(list) {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list) {
    return list.map(s => CartInputTyped.fromStringObj(s));
  }

  toJsonValue() {
    return {
      totalItems: this.totalItems == null ? null : String(this.totalItems),
      shipping: this.shipping == null ? null : String(this.shipping),
      discount: this.discount == null ? null : String(this.discount),
      totalPrice: this.totalPrice == null ? null : String(this.totalPrice),
      notes: this.notes,
    };
  }

  toJSON() { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m) {
    return new CartInputTyped(
      new Dollar(_json.asString(_json.require(m, "totalItems"), "totalItems")),
      new Dollar(_json.asString(_json.require(m, "shipping"), "shipping")),
      new Dollar(_json.asString(_json.require(m, "discount"), "discount")),
      new Dollar(_json.asString(_json.require(m, "totalPrice"), "totalPrice")),
      _json.asString(_json.require(m, "notes"), "notes")
    );
  }

  static fromJSON(text) { return CartInputTyped.fromJsonValue(_json.parse(text)); }

  static toJSONList(items) {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text) {
    const raw = _json.asArray(_json.parse(text), "CartInputTyped");
    return raw.map((e) => CartInputTyped.fromJsonValue(e));
  }

  toString() {
    return `TotalItems=${this.totalItems}, Shipping=${this.shipping}, Discount=${this.discount}, Total Price=${this.totalPrice}, Notes=${this.notes}`;
  }

  equals(other) {
    if (!(other instanceof CartInputTyped)) return false;
    return this.totalItems === other.totalItems
      && this.shipping === other.shipping
      && this.discount === other.discount
      && this.totalPrice === other.totalPrice
      && this.notes === other.notes;
  }
}
